classes/auto_grader/average_grade_no_straddle.php

```
@@ -86,41 +109,30 @@ private function grades_are_close_enough(): bool {
      */
-    public function grades_straddle_class_boundaries(array $gradeclassesadminsetting): bool {
-        $grades = $this->grades_as_percentages();
-        if (empty($grades)) {
-            // No grades found, so we are not applying this rule.
-            return false;
-        }
-
-        $gradeclassesseen = [];
-        foreach ($grades as $grade) {
-            foreach ($gradeclassesadminsetting as $index => $gradeclassboundaries) {
-                $boundarybottom = $gradeclassboundaries[0];
-                $boundarytop = $gradeclassboundaries[1];
-                if ($grade >= $boundarybottom && $grade <= $boundarytop) {
-                    // Grade is within this class.
-                    if (!in_array($index, $gradeclassesseen)) {
-                        $gradeclassesseen[] = $index;
-                    }
-                    if (count($gradeclassesseen) > 1) {
-                        // We have seen more than one grade class, so the grades straddle class boundaries.
-                        return true;
-                    }
-                }
+    public function get_grade_range_index(float $gradepercentage, array $gradeclassesadminsetting): ?int {
+        foreach ($gradeclassesadminsetting as $index => $gradeclassboundaries) {
+            $boundarybottom = $gradeclassboundaries[0];
+            $boundarytop = $gradeclassboundaries[1];
+            $roundeddowngradepercentage = floor($gradepercentage * 100) / 100;
+            // For grade band classification purposes, we round $gradepercentage down to 2 decimal places.
+            // E.g. a percentage grade of 59.999% would be rounded down to 59.99% to determine which band it was in.
+            if ($roundeddowngradepercentage >= $boundarybottom && $roundeddowngradepercentage <= $boundarytop) {
+                // Grade is within this class.
+                return $index;
             }
         }
-        return false;
+        return null;
```

I would say this should throw an exception

---

I didn't want to risk making a submission un-gradeable by throwing an exception.  There is safety in the fact that no auto grade will be generated where null is returned here, and the agreed grade has to be entered manually.  The debug message records that there was an issue.